wfe2/wfe.go

```diff
@@ -1053,12 +1053,88 @@ func (wfe *WebFrontEndImpl) setCORSHeaders(response http.ResponseWriter, request
 }
 
 // KeyRollover allows a user to change their signing key
-func (wfe *WebFrontEndImpl) KeyRollover(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
-	// KeyRollover is a NOP for WFEv2 at the present time. It needs unique JWS
-	// validation compared to other ACME v2 endpoints.
-	wfe.sendError(response, logEvent,
-		probs.ServerInternal("KeyRollover is not presently implemented for ACME v2"), nil)
-	return
+func (wfe *WebFrontEndImpl) KeyRollover(
+	ctx context.Context,
+	logEvent *requestEvent,
+	response http.ResponseWriter,
+	request *http.Request) {
+	// Validate the outer JWS on the key rollover in standard fashion using
+	// validPOSTForAccount
+	outerBody, outerJWS, acct, prob := wfe.validPOSTForAccount(request, ctx, logEvent)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Parse the inner JWS from the validated outer JWS body
+	innerJWS, prob := wfe.parseJWS(outerBody)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Validate the inner JWS as a key rollover request for the outer JWS
+	rolloverRequest, prob := wfe.validKeyRollover(outerJWS, innerJWS, logEvent)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+	newKey := rolloverRequest.NewKey
+
+	// Check that the rollover request's account URL matches the account URL used
+	// to validate the outer JWS
+	accountURL := wfe.jwsKeyIDToAccountID(outerJWS)
+	if rolloverRequest.Account != accountURL {
+		wfe.sendError(response, logEvent, probs.Malformed(
+			fmt.Sprintf("Inner key rollover request specified Account %q, but outer JWS has Key ID %q",
+				rolloverRequest.Account, accountURL)), nil)
+		return
+	}
+
+	// Check that the new key isn't the same as the old key. This would fail as
+	// part of the subsequent `wfe.SA.GetRegistrationByKey` check since the new key
+	// will find the old registration if its equal to the old registration key. We
+	// check new key against old key explicitly to save an RPC round trip and a DB
+	// query for this easy rejection case
+	keysEqual, err := core.PublicKeysEqual(newKey.Key, acct.Key.Key)
+	if err != nil {
+		// This should not happen - both the old and new key have been validated by now
+		wfe.sendError(response, logEvent, probs.ServerInternal("Unable to compare new and old keys"), nil)
+		return
+	}
+	if keysEqual {
+		wfe.sendError(response, logEvent, probs.Malformed(
+			"New key specified by rollover request is the same as the old key"), nil)
+		return
+	}
+
+	// Check that the new key isn't already being used for an existing account
+	if existingAcct, err := wfe.SA.GetRegistrationByKey(ctx, &newKey); err != nil {
+		response.Header().Set("Location", wfe.relativeEndpoint(request, fmt.Sprintf("%s%d", regPath, existingAcct.ID)))
```

Comment:
> also: if we didn't do this we could combine this with the RA.UpdateRegistration call below and reduce the number of RPC round trips. I think we should be able to do a test that the returned error for that is a duplicate error... possibly?

I took a look at the RA/SA `UpdateRegistration` implementations and I don't think it will be trivial to check for whether the update failed because of a unique constraint on the key field failing. I think the SA should probably be updated to catch that case and return a distinct error so we don't leak too much SQL semantics up to the WFE layer. Do you want to create a follow-up issue for this?